### PR TITLE
feat/add coredump debugging script

### PR DESCRIPTION
**Pull request by zahra121281**

### CODE REVIEW

File: gdb_automate.py [hunk 1]
```diff
@@ -43,44 +45,129 @@ def check_port(ssh_conn, password, username):
...
+        so_paths = []
+        for line in readelf_output.stdout.splitlines():
+            match = re.search(r"(/.+\.so(?:\.[\d]+)*)", line)
+            if match:
+                lib =  match.group(1).split('/')[-1]
+                so_paths.append(lib)
+        return so_paths
+
+    except subprocess.CalledProcessError as e:
+        print(f"Error running readelf: {e}")
```
Comment (azharivs) [inline gdb_automate.py]: please remove fstring!

File: gdb_automate.py [hunk 2]
```diff
@@ -43,44 +45,129 @@ def check_port(ssh_conn, password, username):
...
+        lib_path = os.path.join(search_path, lib)
+        if os.path.exists(lib_path):
+            found.append(lib)
+        else:
+            not_found.append(lib)
+    result = False
+    if len(not_found ) > 0 : 
+        print("bellow libraries don't found in the workspace." )
+        for lib in not_found:
+            print(f"  - {lib}")
```
Comment (azharivs) [inline gdb_automate.py]: fstring!

File: gdb_commands/end_command.py [hunk 3]
```diff
@@ -10,27 +10,40 @@
...
+            target_info = gdb.execute('info target', to_string=True)
+            if "Remote serial target" in target_info:
+                self.ip = gdb.execute('show environment IP_ADDRESS', to_string=True).strip().split(' = ' )[1] 
+                self.username = gdb.execute('show environment USERNAME', to_string=True).strip().split(' = ')[1]
+                self.password = gdb.execute('show environment PASSWORD', to_string=True).strip().split(' = ')[1]
+            else:
+                self.ip = None 
+
+        except gdb.error as e:
+            print(f"Error fetching details from target: {e}")
```
Comment (azharivs) [inline gdb_commands/end_command.py]: fstring!

File: gdb_automate.py [hunk 4]
```diff
@@ -43,44 +45,129 @@ def check_port(ssh_conn, password, username):
...
+    
+    for lib in libraries:
+        lib_path = os.path.join(search_path, lib)
+        if os.path.exists(lib_path):
+            found.append(lib)
+        else:
+            not_found.append(lib)
+    result = False
+    if len(not_found ) > 0 : 
+        print("bellow libraries don't found in the workspace." )
```
Comment (azharivs) [inline gdb_automate.py]: Grammar correction:

"Following libraries were not found in workspace."

File: gdb_automate.py [hunk 5]
```diff
@@ -43,44 +45,129 @@ def check_port(ssh_conn, password, username):
...
+    libraries = extract_shared_libraries_from_core(core_dump_path)
+    
+    for lib in libraries:
+        lib_path = os.path.join(search_path, lib)
+        if os.path.exists(lib_path):
+            found.append(lib)
+        else:
+            not_found.append(lib)
+    result = False
+    if len(not_found ) > 0 :
```
Comment (azharivs) [inline gdb_automate.py]: remove extra space before )

File: gdb_automate.py [hunk 6]
```diff
@@ -43,44 +45,129 @@ def check_port(ssh_conn, password, username):
...
+        if os.path.exists(lib_path):
+            found.append(lib)
+        else:
+            not_found.append(lib)
+    result = False
+    if len(not_found ) > 0 : 
+        print("bellow libraries don't found in the workspace." )
+        for lib in not_found:
+            print(f"  - {lib}")
+        cont = input("1- continue with this libraries at workspace\n"
```
Comment (azharivs) [inline gdb_automate.py]: Please enter one option:\n
[1] Continue with whatever libraries found in workspace

File: gdb_automate.py [hunk 7]
```diff
@@ -43,44 +45,129 @@ def check_port(ssh_conn, password, username):
...
+             "2- use libraries in default path in local system.\n"
```
Comment (azharivs) [inline gdb_automate.py]: where is the default path? Can you also print it here?

File: gdb_automate.py [hunk 8]
```diff
@@ -43,44 +45,129 @@ def check_port(ssh_conn, password, username):
...
+             "3- exit program.\n")
+        if cont == "3":
+            print("Exiting the program.")
```
Comment (azharivs) [inline gdb_automate.py]: "Exiting" is enough

File: gdb_commands/bins.py [hunk 9]
```diff
@@ -122,7 +122,7 @@ def invoke(self, arg, from_tty):
         state_manager.bins = []
         PrettyPrinter.print_header("Analysing Bins", width=table_width)
         for ar in state_manager.arenas:
-            print()
+
```
Comment (azharivs) [inline gdb_commands/bins.py]: there is no need for empty line here, the PrettyPrinter is part of the loop block

File: gdb_commands/end_command.py [hunk 10]
```diff
@@ -10,27 +10,40 @@
...
-            ssh_conn.connect()
-            gdbserver_command = 'echo "{}" | sudo -S pkill gdbserver'.format(self.password)
-            if self.username == "root":
-                gdbserver_command = 'pkill gdbserver'.format(self.password)
-            output , error = ssh_conn.run_command(gdbserver_command)
-            
-            print( "killing gdbserver process on remote...")
-            time.sleep( 2) 
+            if self.ip != None : 
+                print("detach from the process " )
```
Comment (azharivs) [inline gdb_commands/end_command.py]: extra space before )

File: gdb_commands/end_command.py [hunk 11]
```diff
@@ -10,27 +10,40 @@
...
-            time.sleep( 2) 
+            if self.ip != None : 
+                print("detach from the process " ) 
+                ssh_conn = SSHConnection(ip=self.ip, username=self.username, password=self.password)
+                gdb.execute("detach")
+                time.sleep(2)            
+                ssh_conn.connect()
+                gdbserver_command = 'echo "{}" | sudo -S pkill gdbserver'.format(self.password)
+                if self.username == "root":
+                    gdbserver_command = 'pkill gdbserver'.format(self.password)
```
Comment (azharivs) [inline gdb_commands/end_command.py]: there should be a placeholder {} for self.password!!

File: gdb_commands/end_command.py [hunk 12]
```diff
@@ -10,27 +10,40 @@
...
+                output , error = ssh_conn.run_command(gdbserver_command)
+                
+                print( "killing gdbserver process on remote...")
```
Comment (azharivs) [inline gdb_commands/end_command.py]: extra space after (

File: gdb_commands/heap.py [hunk 13]
```diff
@@ -52,24 +62,16 @@ def count_memory_usage(self, inferior, heap_start, heap_end):
...
         nxhdr = 0
-        
         while (cur != heap_end):
             num_words = 1
-            memory = inferior.read_memory(
-                cur + state_manager.word_size, num_words * 4)
-
-            word = memory[0:4]
-            value = int.from_bytes(word, byteorder=state_manager.byteorder)
+            value = self.read_memory(cur , num_words)
```
Comment (azharivs) [inline gdb_commands/heap.py]: extra space before ,

File: gdb_commands/heap.py [hunk 14]
```diff
@@ -126,11 +133,22 @@ def extract_heaps(self):
...
+            'p main_arena.top',
+            to_string=True)
+        pattern = r"0x[0-9a-fA-F]+"
+        match = re.search(pattern, top_chunk)
+        hex_address = match.group()
+        value = self.read_memory(int(hex_address , 16 ) , num_words=1)
+        sz = value & ~0x7
+        main_heap.end = str(hex(int(hex_address,16) + sz))
+
+
```
Comment (azharivs) [inline gdb_commands/heap.py]: one empty line is enough

File: gdb_commands/process.py [hunk 15]
```diff
@@ -1,31 +1,41 @@
...
-            num_regions = int(mmap_info['hblks'])
-            total_size = int(mmap_info['hblkhd'])
-            return num_regions, total_size
+            if state_manager.is_process : 
+                mmap_info = gdb.parse_and_eval('mallinfo()')
+                num_regions = int(mmap_info['hblks'])
+                total_size = int(mmap_info['hblkhd'])
+                return num_regions, total_size
+            else : 
+                return "not supported in coredump debugging" , "not supported in coredump debugging"
```
Comment (azharivs) [inline gdb_commands/process.py]: I would do something like:

                print("not supported in coredump debugging")
                return None , None

File: gdb_coredump.py [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+import sys
+# ANSI escape codes for colored text
+YELLOW = '\033[93m'
+RESET = '\033[0m'
+
+
+def main(argv):
+    parser = argparse.ArgumentParser(
+        description='Take the path of coredump file and set the workspace for debugging coredump.')
+    parser.add_argument('-a', '--architecture', type=str, required=True,
```
Comment (azharivs) [inline gdb_coredump.py]: why don't you set the default value right here and not later in the code?
Also provide a few examples in help to show the options for --architecture. A better way is to select from a predefined list of choises as explained in this example:

https://docs.python.org/2/library/argparse.html#choices

File: setup_local.py [hunk 17]
```diff
@@ -20,7 +16,7 @@ def is_package_installed(self, package_name):
         )
         return result.returncode == 0
 
-    def install_gnome_terminal(self):
+    def install_gnome_terminal(self):
```
Comment (azharivs) [inline setup_local.py]: The only change you did are four whitespaces at the end of the line. It is time to install a good python code style plugin so you can auto detect these mistakes.

**Review by azharivs — CHANGES_REQUESTED** [review]

File: gdb_automate.py (hunk 1 above)
Comment (zahra121281) [reply gdb_automate.py]: removed

**Pull request merged**